File: crates/hxd-session/src/frame.rs
//! Transaction framing over a byte stream.
//!
//! The 22-byte header layout (`type`/`trans`/`flag`/`len`/`len2`/`hc`) and
//! all chunk encoding come from `hotline-proto` — the same code gtkhx frames
//! with, so the two ends can never disagree by drift.
//!
//! The read side frames by **`len2` (DataSize), not `len` (TotalSize)** —
//! the distinction only matters for fragmenting senders, but it cost gtkhx a
//! real desync to learn, so the server starts on the right field. No known
//! client fragments requests; a frame whose `len` disagrees with `len2` is
//! rejected rather than half-understood.

use hotline_proto::build::pack_header;
use hotline_proto::wire::ChunkIter;
use hotline_proto::{HL_DATA_HDR_LEN, HL_HDR_LEN};
use tokio::io::{AsyncRead, AsyncReadExt};

/// Hard cap on a transaction's data size — mhxd's `MAX_HOTLINE_PACKET_LEN`.
/// Anything larger is a hostile or broken client.
pub const MAX_FRAME_DATA: u32 = 0x40000;

/// A received transaction. `buf` holds header + body contiguously so
/// `hotline-proto`'s [`ChunkIter`] can walk it in place.
#[derive(Debug)]
pub struct Frame {
    pub ty: u32,
    pub trans: u32,
    pub flag: u32,
    pub hc: u16,
    buf: Vec<u8>,
}

impl Frame {
    /// Iterate the data chunks.
    pub fn chunks(&self) -> ChunkIter<'_> {
        ChunkIter::over_message(&self.buf, self.buf.len())
    }
}

/// Why a read failed.
#[derive(Debug)]
pub enum ReadError {
    /// Clean EOF at a frame boundary — the peer hung up.
    Eof,
    Io(std::io::Error),
    /// The header violates the protocol (oversized, len/len2 mismatch).
    Malformed(String),
}

impl From<std::io::Error> for ReadError {
    fn from(e: std::io::Error) -> Self {
        if e.kind() == std::io::ErrorKind::UnexpectedEof {
            ReadError::Eof
        } else {
            ReadError::Io(e)
        }
    }
}

/// Read one transaction.
pub async fn read_frame<R: AsyncRead + Unpin>(r: &mut R) -> Result<Frame, ReadError> {
    let mut hdr = [0u8; HL_HDR_LEN];
    // Distinguish "EOF before any byte" (clean close) from a torn header.
    if r.read(&mut hdr[..1]).await? == 0 {
        return Err(ReadError::Eof);
    }
    r.read_exact(&mut hdr[1..]).await?;

    let ty = u32::from_be_bytes(hdr[0..4].try_into().unwrap());
    let trans = u32::from_be_bytes(hdr[4..8].try_into().unwrap());
    let flag = u32::from_be_bytes(hdr[8..12].try_into().unwrap());
    let len = u32::from_be_bytes(hdr[12..16].try_into().unwrap());
    let len2 = u32::from_be_bytes(hdr[16..20].try_into().unwrap());
    let hc = u16::from_be_bytes(hdr[20..22].try_into().unwrap());

    if len2 > MAX_FRAME_DATA {
        return Err(ReadError::Malformed(format!(
            "data size {len2:#x} exceeds cap {MAX_FRAME_DATA:#x}"
        )));
    }
    if len != len2 {
        // Clients never fragment; see module docs.
        return Err(ReadError::Malformed(format!(
            "fragmented frame from client (len {len:#x} != len2 {len2:#x})"
        )));
    }

    // len2 counts the body plus the 2-byte `hc` that physically sits at the
    // header's tail; body bytes still to read = len2 - 2.
    let body_len = len2.saturating_sub(2) as usize;
    let mut buf = Vec::with_capacity(HL_HDR_LEN + body_len);
    buf.extend_from_slice(&hdr);
    buf.resize(HL_HDR_LEN + body_len, 0);
    r.read_exact(&mut buf[HL_HDR_LEN..]).await?;

    Ok(Frame {
        ty,
        trans,
        flag,
        hc,
        buf,
    })
}

/// Pack a full outgoing transaction. No chunk-count limit (a user-list
/// reply carries one chunk per user), hence not `pack_message` — but the
/// header bytes come from the same `pack_header` gtkhx uses.
pub fn pack_frame(ty: u32, trans: u32, flag: u32, chunks: &[(u16, Vec<u8>)]) -> Vec<u8> {
    let body_len: usize = chunks.iter().map(|(_, d)| HL_DATA_HDR_LEN + d.len()).sum();
    let mut out = vec![0u8; HL_HDR_LEN];
    pack_header(
        &mut out,
        ty,
        trans,
        flag,
        chunks.len() as u16,
        body_len as u32,
    );
    for (tag, data) in chunks {
        assert!(
            data.len() <= u16::MAX as usize,
            "chunk payload exceeds u16 length"
        );
        out.extend_from_slice(&tag.to_be_bytes());
        out.extend_from_slice(&(data.len() as u16).to_be_bytes());
        out.extend_from_slice(data);
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn frame_roundtrips_through_pack_and_read() {
        let bytes = pack_frame(
            0x0000_006b,
            7,
            0,
            &[(0x0066, b"misha".to_vec()), (0x0068, vec![0x00, 0x91])],
        );
        let mut cur = bytes.as_slice();
        let f = read_frame(&mut cur).await.unwrap();
        assert_eq!((f.ty, f.trans, f.flag, f.hc), (0x6b, 7, 0, 2));
        let chunks: Vec<_> = f.chunks().collect();
        assert_eq!(chunks.len(), 2);
        assert_eq!(chunks[0].tag, 0x0066);
        assert_eq!(chunks[0].data, b"misha");
        assert_eq!(chunks[1].as_uint(), 0x91);
    }

    #[tokio::test]
    async fn empty_body_frame_is_fine() {
        let bytes = pack_frame(0x0001_0000, 3, 0, &[]);
        // len/len2 must be 2 (just the hc tail) for an empty reply.
        assert_eq!(&bytes[12..16], &[0, 0, 0, 2]);
        let mut cur = bytes.as_slice();
        let f = read_frame(&mut cur).await.unwrap();
        assert_eq!(f.chunks().count(), 0);
    }

    #[tokio::test]
    async fn oversized_and_fragmented_frames_are_rejected() {
        let mut bytes = pack_frame(0x69, 1, 0, &[]);
        bytes[16..20].copy_from_slice(&(MAX_FRAME_DATA + 1).to_be_bytes());
        let mut cur = bytes.as_slice();
        assert!(matches!(
            read_frame(&mut cur).await,
            Err(ReadError::Malformed(_))
        ));

        let mut bytes = pack_frame(0x69, 1, 0, &[]);
        bytes[12..16].copy_from_slice(&99u32.to_be_bytes());
        let mut cur = bytes.as_slice();
        assert!(matches!(
            read_frame(&mut cur).await,
            Err(ReadError::Malformed(_))
        ));
    }

    #[tokio::test]
    async fn clean_eof_vs_torn_header() {
        let mut cur: &[u8] = &[];
        assert!(matches!(read_frame(&mut cur).await, Err(ReadError::Eof)));
        let mut cur: &[u8] = &[0x00, 0x01];
        assert!(matches!(read_frame(&mut cur).await, Err(ReadError::Eof)));
    }
}

File: e2e/harness/frame.mjs
/**
 * The legacy transaction frame, written from the wire format rather than
 * from the server's code.
 *
 * That independence is the entire point of this file. Every legacy test
 * in `crates/hxd/tests/` packs with `hxd_session::frame::pack_frame` and
 * parses with `read_frame` — the server's own functions — so a framing
 * bug symmetric between the two is invisible to all of them. Here the
 * bytes are laid out from `docs` and from the 22-byte header itself, so
 * a disagreement shows up as a disagreement.
 *
 * Header, big-endian throughout:
 *
 *     0   type   u32
 *     4   trans  u32
 *     8   flag   u32
 *     12  len    u32   TotalSize
 *     16  len2   u32   DataSize
 *     20  hc     u16   chunk count
 *
 * `len` and `len2` both count the body *plus* the two bytes of `hc`,
 * which is why the body that follows is `len2 - 2` long.
 */

export const HDR_LEN = 22;
export const CHUNK_HDR_LEN = 4;
/** mhxd's `MAX_HOTLINE_PACKET_LEN`. */
export const MAX_FRAME_DATA = 0x40000;

/** `"TRTPHOTL" 0x0001 0x0002`, and the eight bytes back. */
export const CLIENT_MAGIC = Buffer.from('TRTPHOTL\x00\x01\x00\x02', 'binary');
export const SERVER_MAGIC = Buffer.from('TRTP\x00\x00\x00\x00', 'binary');

/** Chunks are `(tag u16, len u16, bytes)`. */
export function pack(type, trans, flag, chunks) {
  const bodyLen = chunks.reduce((n, [, data]) => n + CHUNK_HDR_LEN + data.length, 0);
  const wireLen = bodyLen + 2; // + sizeof(hc)
  const head = Buffer.alloc(HDR_LEN);
  head.writeUInt32BE(type, 0);
  head.writeUInt32BE(trans, 4);
  head.writeUInt32BE(flag, 8);
  head.writeUInt32BE(wireLen, 12);
  head.writeUInt32BE(wireLen, 16);
  head.writeUInt16BE(chunks.length, 20);
  const body = [];
  for (const [tag, data] of chunks) {
    const th = Buffer.alloc(CHUNK_HDR_LEN);
    th.writeUInt16BE(tag, 0);
    th.writeUInt16BE(data.length, 2);
    body.push(th, Buffer.from(data));
  }
  return Buffer.concat([head, ...body]);
}

/** A parsed transaction: the header fields, and its chunks as a list of
 *  `{ tag, data }` in wire order (a tag can repeat — the user list is
 *  one chunk per user). */
export class Frame {
  constructor(type, trans, flag, chunks) {
    this.type = type;
    this.trans = trans;
    this.flag = flag;
    this.chunks = chunks;
  }

  /** The first chunk with this tag, or undefined. */
  get(tag) {
    return this.chunks.find((c) => c.tag === tag)?.data;
  }

  all(tag) {
    return this.chunks.filter((c) => c.tag === tag).map((c) => c.data);
  }
}

/**
 * Accumulate bytes and yield whole transactions.
 *
 * Framing is by **`len2` (DataSize), not `len` (TotalSize)**. The
 * distinction only shows up against a sender that fragments, and it cost
 * gtkhx a real desync to learn, so this reads the field the server
 * reads. A frame whose two lengths disagree is rejected outright rather
 * than half-understood — no known client fragments, so a mismatch is a
 * broken or hostile peer either way.
 */
export class Framer {
  constructor() {
    this.buf = Buffer.alloc(0);
  }

  push(bytes) {
    this.buf = Buffer.concat([this.buf, bytes]);
    const frames = [];
    for (;;) {
      if (this.buf.length < HDR_LEN) break;
      const type = this.buf.readUInt32BE(0);
      const trans = this.buf.readUInt32BE(4);
      const flag = this.buf.readUInt32BE(8);
      const len = this.buf.readUInt32BE(12);
      const len2 = this.buf.readUInt32BE(16);
      const hc = this.buf.readUInt16BE(20);
      if (len2 > MAX_FRAME_DATA) throw new Error(`data size ${len2} exceeds the cap`);
      if (len !== len2) throw new Error(`len ${len} disagrees with len2 ${len2}`);
      const bodyLen = len2 - 2;
      if (this.buf.length < HDR_LEN + bodyLen) break;

      const body = this.buf.subarray(HDR_LEN, HDR_LEN + bodyLen);
      const chunks = [];
      let at = 0;
      for (let i = 0; i < hc; i++) {
        if (at + CHUNK_HDR_LEN > body.length) throw new Error('chunk header runs past the body');
        const tag = body.readUInt16BE(at);
        const size = body.readUInt16BE(at + 2);
        at += CHUNK_HDR_LEN;
        if (at + size > body.length) throw new Error(`chunk ${tag} runs past the body`);
        chunks.push({ tag, data: Buffer.from(body.subarray(at, at + size)) });
        at += size;
      }
      frames.push(new Frame(type, trans, flag, chunks));
      this.buf = this.buf.subarray(HDR_LEN + bodyLen);
    }
    return frames;
  }
}
